File: 2078-two-furthest-houses-with-different-colors/2078-two-furthest-houses-with-different-colors.go
func maxDistance(colors []int) int {
	n := len(colors)
	left, right := 0, n-1
	ans := -1
	for left <= right {
		if colors[left] != colors[right] {
			ans = right - left
		}

		// this is so f greedy
		if colors[left] != colors[n-1] || colors[right] != colors[0] {
			ans = right - 0 // "(n - 1) - left" is also valid
		}

        if ans != -1 { return ans }

		left++
		right--
	}

	return 1
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

// array, greedy
// time: O(n)
// space: O(1)

File: 2078-two-furthest-houses-with-different-colors/2078-two-furthest-houses-with-different-colors.py
class Solution:
    def maxDistance(self, colors: List[int]) -> int:
        n = len(colors)
        ans = -1

        for i in range(n - 1, -1, -1):
            if colors[i] != colors[0]:
                ans = max(ans, i - 0)
                break

        for i in range(n):
            if colors[i] != colors[n-1]:
                ans = max(ans, n - 1 - i)
        
        return ans

# array, greedy
# time: O(n)
# space: O(1)